### ISSUE TITLE
Creating code samples from OpenAPI schema data

### ISSUE THREAD

**Issue by zeke** [octokit/routes#484]
In #168, a new OpenAPI schema landed in this project, and will pave the way for a future where all the octokits have a shared schema! ❤️ 

Today I was inspecting the contents of the [openapi/api.github.com/operations](https://github.com/octokit/routes/tree/261c763e51ea4d327fedc51cde182d2ffec65708/openapi/api.github.com/operations) directory, and noticed that `operationId` values are all in snake-case format:

```
"operationId": "pulls-create-comment-reply"
```

It looks like these values are currently created by joining the `scope` and the `endpoint.idName` with a hyphen:

https://github.com/octokit/routes/blob/261c763e51ea4d327fedc51cde182d2ffec65708/lib/openapi.js#L193

I also noticed that there's already some code in this project for creating example code with `toShellExample` and `toJsExample`:

https://github.com/octokit/routes/blob/261c763e51ea4d327fedc51cde182d2ffec65708/lib/openapi.js#L261)

This leads me to a few questions:

- Do the generated code examples work yet? Based on the current implementation of `toJsExample` my guess is no:

https://github.com/octokit/routes/blob/261c763e51ea4d327fedc51cde182d2ffec65708/lib/openapi.js#L269

- Can we switch to using `operationId` values that more closely match the method names octokit clients will use? For example, I'd suggest that `pulls-create-comment-reply` should instead be `pulls.createCommentReply`

I'm happy to help out on this, but wanted to open this issue first to discuss and get a sense of where things are with this effort before jumping to an implementation.

cc @gr2m @d11n

**Comment by gr2m** [comment]
`operationId` is language agnostic, I thought hyphens would be the easiest to work with. The respective language can then turn it into what ever feels native, such as camelCase for JavaScript.

I had to prefix `operationId` with the scope because operationIds need to be globally unique. Without the scope prefix there would be conflicts

The code examples work for JS should work, you can see it rendered at https://octokit-routes-openapi.now.sh.

The challenge with that approach is that all SDKs need to have the new endpoints implemented before the REST API docs gets published. And even then it would be helpful to know which version of the SDK introduced the new method.

Maybe the code examples should live in the repositories of the respective SDKs and then be pulled in from there into the central developer docs?

**Comment by d11n** [comment]
yeah...and in ruby, for example, PascalCase is used for modules/classes and snake_case for methods...so `pulls.createCommentReply` in octokit.rb would actually be:
```ruby
Octokit::Client::Pulls.create_comment_reply
```
although currently it's actually:
```ruby
Octokit::Client::PullRequests.create_pull_request_comment_reply
Octokit::Client::PullRequests.create_pull_reply   #alias
Octokit::Client::PullRequests.create_review_reply #alias
```
for reference, the current non-`octokit/routes`-powered octokit.rb: https://github.com/octokit/octokit.rb/blob/master/lib/octokit/client/pull_requests.rb#L218-L236

**Comment by d11n** [comment]
and i like the idea of sdk maintainers being responsible for the code that generates the examples...but perhaps it stays in this repo like:

```
./lib/endpoint/code-samples/curl.shell.js
                            gh3.js.js
                            github.js.js
                            github-client.coffee.js
                            octokit.net.js
                            octokit.rb.js
                            octokit-rest.js.js
                            octonode.js.js
                            ...
```

and the function signatures could be something like:

```javascript
function toExample ({ route, scope, baseUrl }) {
  // insert unholy js written by non-jsers ;P
  return {
    sdk: {
      name,
      url,
      ...
    },
    infoString, // code fence info string
    // ^ https://github.github.com/gfm/#example-112
    //   https://github.com/github/linguist/blob/master/vendor/README.md
    codeString
  }
}
```

**Comment by zeke** [comment]
> The code examples work for JS should work

It looks like they're all currently rendering as `octokit.{scope}.get)(...)`. Here's an example:

![Screen Shot 2019-08-07 at 10 43 42 AM](https://user-images.githubusercontent.com/2289/62645091-6197fd00-b900-11e9-800d-8ce7d2dbd1b6.png)

---

> `operationId` is language agnostic, I thought hyphens would be the easiest to work with. The respective language can then turn it into what ever feels native, such as camelCase for JavaScript.

Are scopes always a single word? If so, I suppose it's easy enough to assemble a method by extracting the first part before the hyphen. My suggestion is not about a specific capitalization preference, but more a clear way to differentiate the scope from the function name. I feel it's easier to derive `Pulls.create_comment_reply` from `pulls.createCommentReply` because you have a `.` separating the two.

**Comment by gr2m** [comment]
> It looks like they're all currently rendering as `octokit.{scope}.get)(...)`

Ooopsiedooopsie

> Are scopes always a single word

Not quite, there is `oauth_authorizations` for example. 

> My suggestion is not about a specific capitalization preference, but more a clear way to differentiate the scope from the function name

Sure why not, I don't like the current mix of underscores and dashes
